osquery_rules/linux/osquery_logins_non_office.py

```diff
@@ -0,0 +1,31 @@
+import ipaddress
+
+# Note: The IP Network below is only an example
+OFFICE_NETWORK = ipaddress.ip_network('192.0.1.0/24')
+
+
+def rule(event):
+    if event['action'] != 'added':
```

Comment:
In general, I've been trying to do safe lookups (either with `.get()` or by first checking for presence of the key). Is this field guaranteed to always be present for every single `Osquery.Differential` log processed?

Reply:
I want @rleighton and @kostaspap opinions on this. My mentality is that for top-level keys, we should just do direct key access without `.get()`.

Reply:
I agree, and that's how it used to be for both the cloud security and the log analysis data pipelines. But some implementation details were changed in the log analysis pipeline and this is no longer the case. My understanding is that while some are required, the majority of top level keys are not required and so not guaranteed to be present. So if the `action` key is ever not present, this will raise a key error exception.

Reply:
There's no guarantees that *all* the top level fields will exist. If they were sent from upstream yes, but if they were not they will not be present. 

In this particular case though, `action` will always be there - OsQuery always includes this field. However I think it's better just to follow one specific pattern for accessing keys throughout all our rules.

Reply:
Btw, if we think this is becoming uncomfortable to use we could wrap the `event` with something like: 
```
class Forgive:
  def __init__(self, value = None):
    self.value = value
  def __getitem__(self, name):
    if self.value is None:
      return Forgive()
    try:
      return Forgive(self.value.__getitem__(name))
    except (KeyError, AttributeError):
      return Forgive()
  def get(self, default = None):
    return default if self.value is None else self.value

data = {'actor':{'username': 'Joe'}}
print(Forgive(data)['actor']['username'].get('default1'))
print(Forgive(data)['actor']['address'].get('default2'))
```
My only concern with this is that it will not have the default behavior of a Python dictionary... We can chat about this offline.